### PR TITLE
🔀 :: (#194) 유저 랭크 날짜별 정보 Job

### LINKED ISSUE #194: UserRankJob 생성
### Describe

_No response_

### Work

- [x] Job, Step 설계
- [x] Batch 처리
- [x] 1시간마다 업데이트

### Additional

_No response_

### PR DESCRIPTION
**Pull request by parksangwoo1617**
close #194 

코드 많이 못생겼는데 리펙토링은 시간 남으면 하고 안 남으면 런칭하고 할게요
프로시저 좀 바뀌었으니까 봐주세요

### 저장 프로시저 
**유저 랭크 날짜, 학교별로 조회하는 프로시저**
> DELIMITER $$
CREATE PROCEDURE SELECT_USER_RANK_BY_SCHOOL(IN date_type_in INT)
BEGIN
    SELECT user.name,
           user.school_id,
           user.grade,
           user.class                                                                                                  class_num,
           user.profile_image_url,
           user.id user_id,
           IFNULL(SUM(exercise_analysis.walk_count), 0)                                                             AS walk_count,
           DENSE_RANK() OVER (PARTITION BY user.school_id ORDER BY SUM(exercise_analysis.walk_count) DESC) AS ranking
    FROM user
             LEFT JOIN exercise_analysis ON exercise_analysis.user_id = user.id
    WHERE DATEDIFF(NOW(), exercise_analysis.date) < date_type_in
    GROUP BY user.id, user.name, user.school_id, user.profile_image_url, user.class, user.grade;
END$$
DELIMITER ;

**유저 랭크 날짜, 반별로 조회하는 프로시저**

> DELIMITER $$
CREATE PROCEDURE SELECT_USER_RANK_BY_CLASS(IN date_type_in INT)
BEGIN
    SELECT user.name,
           user.school_id,
           user.grade,
           user.class                                                                                                               class_num,
           user.profile_image_url,
           user.id user_id,
           IFNULL(SUM(exercise_analysis.walk_count), 0)                                                                          AS walk_count,
           DENSE_RANK() OVER (PARTITION BY user.school_id + user.class ORDER BY SUM(exercise_analysis.walk_count) DESC) AS ranking
    FROM user
             LEFT JOIN exercise_analysis ON exercise_analysis.user_id = user.id
    WHERE DATEDIFF(NOW(), exercise_analysis.date) < date_type_in
    GROUP BY user.id, user.name, user.school_id, user.profile_image_url, user.class, user.grade;
END$$
DELIMITER ;

**유저 랭크 저장 프로시저**

> DELIMITER $$
CREATE PROCEDURE SAVE_USER_RANK(IN user_id_in VARCHAR(255),
                                IN created_at_in DATE,
                                IN date_type_in ENUM ('WEEK', 'MONTH'),
                                IN scope_type_in ENUM ('SCHOOL', 'CLASS'),
                                IN school_id_in VARCHAR(7),
                                IN name_in VARCHAR(10),
                                IN grade_in INT,
                                IN class_num_in INT,
                                IN profile_image_url_in VARCHAR(255),
                                IN ranking_in INT,
                                IN walk_count_in INT)
BEGIN

    IF (SELECT count(*)
        FROM user_rank
        WHERE user_id = user_id_in
          AND created_at = created_at_in
          AND date_type = date_type_in
          AND scope_type = scope_type_in) = 0 THEN
        INSERT INTO user_rank (user_id, created_at, date_type, scope_type, school_id, name, grade, class_num,
                               profile_image_url, ranking, walk_count)
        VALUES (user_id_in, created_at_in, date_type_in, scope_type_in, school_id_in, name_in, grade_in, class_num_in,
                profile_image_url_in, ranking_in, walk_count_in);
    ELSE
        UPDATE user_rank
        SET ranking    = ranking_in,
            walk_count = walk_count_in
        WHERE account_id = user_id_in
          AND created_at = created_at_in
          AND date_type = date_type_in
          AND scope_type = scope_type_in;
    END IF;

END $$
DELIMITER ;

### CODE REVIEW

**Comment by softpeanut** [comment]
SchoolRank와 관련된 것은 새로운 이슈(#198)에서 수정하도록 하겠습니다 !

File: src/main/java/com/walkhub/walkhub/domain/rank/domain/SchoolRankId.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import javax.persistence.Column;
+import java.io.Serializable;
+import java.time.LocalDateTime;
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+public class SchoolRankId implements Serializable {
+    @Column(length = 7)
+    private String agencyCode;
```
Comment (eogus2513) [inline src/main/java/com/walkhub/walkhub/domain/rank/domain/SchoolRankId.java:17]: 선배님 저희 agencyCode를 school_id로 다 수정해서 다시 해야될거 같아요..

File: src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRank.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    private String profileImageUrl;
+
+    @Column(nullable = false)
+    private Integer walkCount;
+
+    @Column(columnDefinition = "TINYINT", nullable = false)
+    private Integer ranking;
+
+    @Column(length = 7)
+    private String agencyCode;
```
Comment (eogus2513) [inline src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRank.java:43]: 여기도..

File: src/main/java/com/walkhub/walkhub/domain/rank/presentation/dto/response/UserRankInfo.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.Getter;
+
+@Getter
+@Builder
+public class UserRankInfo {
+    private String accountId;
+
+    private String name;
+
+    private String agencyCode;
```
Comment (eogus2513) [inline src/main/java/com/walkhub/walkhub/domain/rank/presentation/dto/response/UserRankInfo.java:13]: 여기도..

**Review by eogus2513 — APPROVED** [review]

File: src/main/java/com/walkhub/walkhub/domain/rank/domain/SchoolRankId.java (hunk 1 above)
Comment (parksangwoo1617) [reply src/main/java/com/walkhub/walkhub/domain/rank/domain/SchoolRankId.java:17]: 아이고 다 수정하겠습니다

File: src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRankId.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import java.time.LocalDate;
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@EqualsAndHashCode
+public class UserRankId implements Serializable {
+    private String userId;
+
+    private LocalDate createdAt;
```
Comment (eogus2513) [inline src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRankId.java:15]: 여기는 LocalDate로 한 이유가 있을까요?

File: src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRankId.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.io.Serializable;
+import java.time.LocalDate;
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@EqualsAndHashCode
+public class UserRankId implements Serializable {
+    private String userId;
```
[the remark this answers is not in the document]
Comment (parksangwoo1617) [reply src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRankId.java:13]: 넵

File: src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRank.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    private String profileImageUrl;
+
+    @Column(nullable = false)
+    private Integer walkCount;
+
+    @Column(columnDefinition = "TINYINT", nullable = false)
+    private Integer ranking;
+
+    private String schoolId;
```
[the remark this answers is not in the document]
Comment (parksangwoo1617) [reply src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRank.java:42]: schoolId요?

[the remark this answers is not in the document]
Comment (eogus2513) [reply src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRank.java:42]: 넵

File: src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRankId.java (hunk 4 above)
Comment (eogus2513) [reply src/main/java/com/walkhub/walkhub/domain/rank/domain/UserRankId.java:15]: 아하 넵!

**Review by glay415 — APPROVED** [review]

**Review by lyutvs — APPROVED** [review]

File: src/main/java/com/walkhub/walkhub/domain/rank/job/UserRankJob.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    @Bean
+    @StepScope
+    public StoredProcedureItemReader<UserRankInfo> weeklyUserSchoolRankReader(@Value("#{jobParameters[stepKey]}") Integer type) {
+        return callProcedure("weeklyUserSchoolRankReader", "SELECT_USER_RANK_BY_SCHOOL", 7);
+    }
+
+    @Bean
+    @StepScope
+    public StoredProcedureItemReader<UserRankInfo> monthlyUserSchoolRankReader(@Value("#{jobParameters[stepKey]}") Integer type) {
+        return callProcedure("monthlyUserSchoolRankReader", "SELECT_USER_RANK_BY_SCHOOL", 31);
```
Comment (softpeanut) [inline src/main/java/com/walkhub/walkhub/domain/rank/job/UserRankJob.java:101]: 이거 31일이 아니라 28일로 통일했던 것 같아요

Comment (parksangwoo1617) [reply src/main/java/com/walkhub/walkhub/domain/rank/job/UserRankJob.java]: 변경했습니다

**Pull request merged**